### PR TITLE
[GPU] Allocate reorder & permute's output memory for bounded dynamic shape

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
 - Allow reorder & permute to allocate memory at build time for bounded dynamic shape

### Tickets:
 - 116421

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/reorder.cpp [hunk 1]
```diff
@@ -210,9 +210,10 @@ reorder_inst::typed_primitive_inst(network& network) : parent(network) {
 }
 
 reorder_inst::typed_primitive_inst(network& network, reorder_node const& node)
-    : parent(network, node, (!node.can_be_optimized() && node.get_output_layout().is_static()) ? true : false)
+    : parent(network, node, node.can_be_optimized() ? false : node.get_output_layout().is_static() ?
+                                                       true : node.get_output_layout().has_upper_bound() ? true : false)
     , _req_reinterpr(node.requires_reinterpret()) {
-    if (node.can_be_optimized())
+        if (node.can_be_optimized())
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/src/graph/reorder.cpp:216]: nit: why is there an extra tab here?

**Review by andrew-k-park — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/reorder.cpp (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/reorder.cpp]: Fixed, thanks!

**Review by e-ddykim — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/permute.cpp [hunk 2]
```diff
@@ -108,7 +108,8 @@ std::string permute_inst::to_string(permute_node const& node) {
 }
 
 permute_inst::typed_primitive_inst(network& network, permute_node const& node) :
-        parent(network, node, (!node.can_be_optimized() && node.get_output_layout().is_static()) ? true : false) {
+        parent(network, node, node.can_be_optimized() ? false : node.get_output_layout().is_static() ?
+                                                       true : node.get_output_layout().has_upper_bound() ? true : false) {
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/permute.cpp:112]: How about reducing ternary operators count? Just to make it more readable...
```suggestion
        parent(network, node, node.can_be_optimized() ? false
                                                      : node.get_output_layout().is_static() || node.get_output_layout().has_upper_bound()) {
```
Or
```suggestion
        !node.can_be_optimized() && (node.get_output_layout().is_static() || node.get_output_layout().has_upper_bound()) {
```

**Review by sshlyapn — APPROVED** [review]
Looks good to me, just left minor suggestion

File: src/plugins/intel_gpu/src/graph/permute.cpp (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/permute.cpp]: Fixed as commented, thanks.

**Pull request merged**